Avi Networks Ansible modules.

grastogi23 opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
avi

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0
  config file = /home/grastogi/.ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
New Avi Ansible modules. They provide full configuration and API support for Avi Vantage Solution. The modules honor impotency and depend on pip package avisdk.
Added code to
Dependencies: avisdk, requests
Avi Version: 16.3.4
Module maintainers: @grastogi23, @ericsysmin, @khaltore
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

<!-- Paste verbatim command output below, e.g. before and after your change -->
```

```

Commit:
Avi Networks Ansible modules.
Avi Version: 16.3.4

gundalow commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,189 @@
+    http_method:
+        description:
+            - Allowed HTTP methods for RESTful services and are supported by
+            Avi Controller.
```
This needs to be a single line, it's currently invalid YAML and causing the following shippable error.
```
2017-01-18 19:04:21 ============================================================================
2017-01-18 19:04:21 lib/ansible/modules/network/avi/avi_api_session.py
2017-01-18 19:04:21 ============================================================================
2017-01-18 19:04:21 TRACE:
2017-01-18 19:04:21     while scanning a simple key
2017-01-18 19:04:21       in "<string>", line 15, column 13:
2017-01-18 19:04:21                     Avi Controller.
2017-01-18 19:04:21                     ^
2017-01-18 19:04:21     could not find expected ':'
2017-01-18 19:04:21       in "avi_api_session.DOCUMENTATION", line 56, column 9:
2017-01-18 19:04:21                 choices: ["get", "put", "post",  ... 
2017-01-18 19:04:21                 ^
2017-01-18 19:04:21 ERROR: DOCUMENTATION is not valid YAML. Line 56 column 9
```

gundalow commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,189 @@
+        choices: ["get", "put", "post", "patch", "delete"]
+        required: true
+    data:
+        description:
```
All `description:` MUST be full sentences (capital letter and full stop)

gundalow commented on lib/ansible/module_utils/avi.py:
```diff
@@ -0,0 +1,37 @@
+#
+# Created on December 12, 2016
+# @author: Gaurav Rastogi (grastogi@avinetworks.com)
+# module_check: not supported
```
Does this module_check make sense in module_utils?

gundalow commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,189 @@
+    if not HAS_AVI:
+        return module.fail_json(msg=(
+            'Avi API SDK is not installed. '
+            'Pl. use sudo pip install --upgrade avisdk to install Avi SDK. '
```
'Use "sudo pip install --upgrade avisdk" to install Avi SDK'

gundalow reviewed (changes requested):
needs_revision

I'll review more later in the week

jctanner commented:
bot_status

Commit:
Fixed Review comments
1. Changed description to be full sentences
2. Fixed Pep8 warnings.
3. Fixed comments and descriptions.

gundalow commented:
FYI Maintainers have been set https://github.com/ansible/ansibullbot/pull/280

bcoca commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,187 @@
+            - HTTP body in YAML format.
+    data_json:
+        description:
+            - HTTP body in JSON format.
```
you really dont need 2 datas here, a type='json' will take both a data structure or json doc

grastogi23 replied:
Yes, the input could be a json blob or a dictionary based data. Current module does not auto convert between those.

gundalow replied:
It would be good to combine these before the module gets released to avoid changing the module interface after release.

bcoca commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,187 @@
+
+short_description: API Module for invoking any Avi API.
+description:
+    - This module is used to configure Pool object
```
this is the long description, it does not say what Pool or from where

bcoca commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,187 @@
+module: avi_api
+author: Gaurav Rastogi (grastogi@avinetworks.com)
+
+short_description: API Module for invoking any Avi API.
```
too many apis

bcoca commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,187 @@
+        changed = not avi_obj_cmp(new_obj, existing_obj)
+    return ansible_return(module, rsp, changed, req=data)
+
+
```
the module in general is just a thin api wrapper, not really that useful for users, I'm not against having an 'expert mode' but this is in addition to actual options usage.

grastogi23 replied:
Actually, this is the only option when you want to use HTTP patch semantics where you don't have the full object. This is actively used. I can provide references to actual field use cases if it would help.

bcoca commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,187 @@
+    if not HAS_AVI:
+        return module.fail_json(msg=(
+            'Avi API SDK is not installed. '
+            'Use "sudo pip install --upgrade avisdk" to install Avi SDK. '
```
we dont normally give exact commands as user environments vary too much, normally giving the url and 'you need to install the avisdk library' should be enough.

bcoca commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,187 @@
+        http_method=dict(required=True,
+                         choices=['get', 'put', 'post', 'patch',
+                                  'delete']),
+        path=dict(required=True),
```
path does not seem to be documented

grastogi23 replied:
fixed

bcoca commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,187 @@
+obj:
+    description: Avi REST resource
+    returned: success, changed
+    type: dict
```
and this shows problem with such generic module, user MUST know api to know what is returned

grastogi23 replied:
No, this is not a problem. User needs to know the API regardless as they need to know the data format. Just having description of high level field names is not enough.

 This is a module that is a proxy for calling all Avi REST API which do not have an object behind them like GET analytics, status etc

bcoca commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,187 @@
+HAS_AVI = True
+try:
+    from avi.sdk.avi_api import ApiSession, ObjectNotFound
+    from avi.sdk.utils.ansible_utils import (
```
im guessing this is supposed to be ansible.module_utils.avi as that is where you define avi_common_argument_spec

grastogi23 replied:
This is a workaround because I would like to use the same module in the avi roles in ansible galaxy. Using ansible.module_utils is not an option as it is not packaged with the roles.

gundalow replied:
If these modules get into Ansible then is their any reason to maintain a copy in Galaxy? Having two sources for modules sounds like it could cause confusion for your modules.

abadger replied:
If you'd like to test on devel (will be 2.3), we added a feature a few weeks ago to allow module_utils to be shipped via roles.  You can put module_utils into a directory named module_utils in the role if you're using that code.

bcoca commented on lib/ansible/modules/network/avi/avi_pool.py:
```diff
@@ -0,0 +1,338 @@
+HAS_AVI = True
+try:
+    from avi.sdk.utils.ansible_utils import (
+        avi_ansible_api, avi_common_argument_spec)
```
we normally want only one module to review, as the same issues are normally repeated, like here

since the first module was too generic, i'm going to continue reviewing this one, in case different patterns arise

bcoca commented on lib/ansible/modules/network/avi/avi_pool.py:
```diff
@@ -0,0 +1,338 @@
+
+short_description: Pool Configuration
+description:
+    - This module is used to configure Pool object
```
not sure why the description is same as previous module

grastogi23 replied:
fixed

bcoca commented on lib/ansible/modules/network/avi/avi_pool.py:
```diff
@@ -0,0 +1,338 @@
+        description:
+            - Server reselect configuration for http requests. 
+    servers:
+        description:
```
some of the options dont have very helpful descriptions:

'A list of servers for the pool to direct trafic to' ... also a note to see the format of the server in the examples would be helpful

used this one as an example, but several of the descriptions could be made clearer

bcoca commented on lib/ansible/modules/network/avi/avi_pool.py:
```diff
@@ -0,0 +1,338 @@
+            'Avi API SDK is not installed. '
+            'Use "sudo pip install --upgrade avisdk" to install Avi SDK. '
+            'For more details visit https://github.com/avinetworks/sdk.'))
+    return avi_ansible_api(module, 'pool',
```
no error handling? we normally try to give 'nice' error messages to user, detailing the action attempted and giving as much information possible to assist them in fixing the issue. i.e `'failed to get current status of Pool: %s' % to_native(e)` when getting an exception while trying to contact the api to get a current status.

grastogi23 replied:
Error handling is done inside the avi_ansible_api and it does provide error messages not just within the ansible module but also from the backend server.

bcoca reviewed (changes requested):
only reviewed the first 2, most of the issues are repeated and i'm assuming reviewing more is just going to show the same.

Commit:
1. Fixed descriptions and messages as per review comments.
2. Added descriptions for the missing parameters.

grastogi23 commented:
I have updated the pull request in response to the comments and and feedback from @bcoca

gundalow commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,192 @@
+            - Query parameters passed to the HTTP API.
+    path:
+        description:
+            - Path for Avi API resource. For example, "path: virtualservice" will translate to api/virtualserivce.
```
`Path for Avi API resource. For example, C(path: virtualservice) will translate to C(api/virtualserivce).`

The ": " is causing a YAML parsing error in Shippable

Commit:
Fixed the shippable break due to the incorrect description format

thaumos commented:
bot_status

thaumos commented:
@grastogi23, could you push a new commit to get a build kicked off?

thaumos reviewed (changes requested):
@grastogi23 for the next commit, could you remove one of the extra modules in the PR?  As mentioned, a module to create the namespace would be preferred.

abadger commented on lib/ansible/module_utils/avi.py:
```diff
@@ -0,0 +1,36 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
```
Just noticed the license here.  Generally, module_utils files are BSD licensed so that people writing custom modules can feel that they are okay to license their custom modules as they choose.

grastogi23 replied:
Hi 
This was a per suggestion from @thaumos . Please let me know if this needs to be changed.

thaumos replied:
Let's change it to BSD.

Commit:
Removed the extra modules so that there is a single module for the first commit

grastogi23 commented:
@thaumos I have removed all modules except the avi_api_session.

Commit:
Updated license to BSD based on review comments

grastogi23 commented:
done.

On Mon, Feb 6, 2017 at 9:35 PM, Dylan Silva <notifications@github.com>
wrote:

> *@thaumos* requested changes on this pull request.
> ------------------------------
>
> In lib/ansible/module_utils/avi.py
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20416716>
> :
>
> > +#
> +#
> +# This file is part of Ansible
> +#
> +# Ansible is free software: you can redistribute it and/or modify
> +# it under the terms of the GNU General Public License as published by
> +# the Free Software Foundation, either version 3 of the License, or
> +# (at your option) any later version.
> +#
> +# Ansible is distributed in the hope that it will be useful,
> +# but WITHOUT ANY WARRANTY; without even the implied warranty of
> +# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
> +# GNU General Public License for more details.
> +#
> +# You should have received a copy of the GNU General Public License
> +# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
>
> Let's change it to BSD.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20416716>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADrBf7I5uSQCoYUZ2XJVI6QAA1N80VDdks5raAK_gaJpZM4LnQxe>
> .
>



-- 
Gaurav Rastogi

grastogi AT avinetworks.com
Avi Networks

gundalow commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,192 @@
+#
+# @author: Gaurav Rastogi (grastogi@avinetworks.com) GitHub ID: grastogi23
+#
+# module_check: not supported
```
What would be needed to support check mode? Idempotent modules is one of Ansible's key features

grastogi23 replied:
It does support impotency fully. However, it does not support module_check as a dry-run because of limitation on Avi API which do not have a dry-run option. Checking will alter state on the Avi controllers.

gundalow commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,192 @@
+module: avi_api
+author: Gaurav Rastogi (grastogi@avinetworks.com)
+
+short_description: Avi API Module.
```
No fullstop on `short_description:`

gundalow commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,192 @@
+            - Path for Avi API resource. For example, C(path: virtualservice) will translate to C(api/virtualserivce).
+    timeout:
+        description:
+            - Timeout for Avi API calls.
```
`Timeout (in seconds) for Avi API calls.` (assuming it is seconds)

gundalow commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,192 @@
+'''
+
+EXAMPLES = '''
+# Get Pool Information using avi_api_session
```
Ansible best practice is to specify `-name:`. This allows people to copy and paste the examples to use in their own playbooks. Therefore it's clearer to write this as:

```
- name: Get Pool Information using avi_api_session
  avi_api_session:
    controller: "{{ controller }}"
...
```

This applies all examples.

gundalow commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,192 @@
+        params=dict(type='dict'),
+        data=dict(type='dict'),
+        data_json=dict(type='str'),
+        timeout=dict(default=60)
```
`, type='int'`

gundalow commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,192 @@
+        tenant_uuid=tenant_uuid)
+
+    tenant = module.params.get('tenant', '')
+    timeout = int(module.params.get('timeout', 60))
```
You've specified the default here and in `argument_specs` which doesn't seem right

gundalow commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,192 @@
+    params:
+        description:
+            - Query parameters passed to the HTTP API.
+    path:
```
In `argument_specs` you've specified that `path` is required, so that needs documenting here.

gundalow commented on lib/ansible/utils/module_docs_fragments/avi.py:
```diff
@@ -0,0 +1,45 @@
+            - Password of Avi user in Avi controller. The default value is the environment variable C(AVI_PASSWORD).
+    tenant:
+        description:
+            - Tenant for the operations
```
Missing fullstop.

gundalow commented on lib/ansible/utils/module_docs_fragments/avi.py:
```diff
@@ -0,0 +1,45 @@
+        default: admin
+    tenant_uuid:
+        description:
+            - Tenant uuid for the operations
```
Missing fullstop.

gundalow commented on lib/ansible/modules/network/avi/avi_api_session.py:
```diff
@@ -0,0 +1,192 @@
+
+short_description: Avi API Module.
+description:
+    - This module can be used for calling any resources defined in Avi REST API. This module is useful for invoking HTTP Patch methods and accessing resources that do not have an REST object associated with them.
```
Please split into under 160 character length lines

Commit:
updated comments based on review feedback

grastogi23 commented:
Oh! that is awesome!! let me try that out and then submit the an updated PR
where I will move utils stuff into the ansible itself.

thanks
Gaurav


On Tue, Feb 7, 2017 at 1:48 PM, Toshio Kuratomi <notifications@github.com>
wrote:

> *@abadger* commented on this pull request.
> ------------------------------
>
> In lib/ansible/modules/network/avi/avi_api_session.py
> <https://github.com/ansible/ansible/pull/20415>:
>
> > +# GNU General Public License for more details.
> +#
> +# You should have received a copy of the GNU General Public License
> +# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
> +#
> +"""
> +
> +import json
> +import time
> +from ansible.module_utils.basic import AnsibleModule
> +from copy import deepcopy
> +
> +HAS_AVI = True
> +try:
> +    from avi.sdk.avi_api import ApiSession, ObjectNotFound
> +    from avi.sdk.utils.ansible_utils import (
>
> If you'd like to test on devel (will be 2.3), we added a feature a few
> weeks ago to allow module_utils to be shipped via roles. You can put
> module_utils into a directory named module_utils in the role if you're
> using that code.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/20415>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADrBf66u3ClaSdkbmfBRnPRKkv8nKcWUks5raOafgaJpZM4LnQxe>
> .
>



-- 
Gaurav Rastogi

grastogi AT avinetworks.com
Avi Networks

Commit:
Refactored code to handle POST and PUT scenarios where playbook does not need to check whether object is present.
Moved ansible helper utilities to module_utils as now roles can be patched with module_utils as well.

grastogi23 commented:
hi
I have updated the PR with following updates.

thanks
Gaurav

On Tue, Feb 7, 2017 at 3:27 AM, John R Barker <notifications@github.com>
wrote:

> *@gundalow* requested changes on this pull request.
> ------------------------------
>
> In lib/ansible/modules/network/avi/avi_api_session.py
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20482258>
> :
>
> > @@ -0,0 +1,192 @@
> +#!/usr/bin/python
> +"""
> +# Created on Aug 12, 2016
> +#
> +# @author: Gaurav Rastogi (grastogi@avinetworks.com) GitHub ID: grastogi23
> +#
> +# module_check: not supported
>
> What would be needed to support check mode? Idempotent modules is one of
> Ansible's key features
> ------------------------------
>
> In lib/ansible/modules/network/avi/avi_api_session.py
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20482258>
> :
>
> > +# GNU General Public License for more details.
> +#
> +# You should have received a copy of the GNU General Public License
> +# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
> +#
> +"""
> +
> +import json
> +import time
> +from ansible.module_utils.basic import AnsibleModule
> +from copy import deepcopy
> +
> +HAS_AVI = True
> +try:
> +    from avi.sdk.avi_api import ApiSession, ObjectNotFound
> +    from avi.sdk.utils.ansible_utils import (
>
> If these modules get into Ansible then is their any reason to maintain a
> copy in Galaxy? Having two sources for modules sounds like it could cause
> confusion for your modules.
>

@grastogi: only avi.sdk.avi_api will remain outside as it is a common Api
module. Everything else I will move into ansible_utils once current modules
are released with Ansible. These are very critical modules so it is better
to have a single deterministic copy in customer installations till avi
modules are fully integrated and released with ansible.


> ------------------------------
>
> In lib/ansible/modules/network/avi/avi_api_session.py
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20482258>
> :
>
> > +HAS_AVI = True
> +try:
> +    from avi.sdk.avi_api import ApiSession, ObjectNotFound
> +    from avi.sdk.utils.ansible_utils import (
> +        ansible_return, avi_obj_cmp, cleanup_absent_fields,
> +        avi_common_argument_spec)
> +except ImportError:
> +    HAS_AVI = False
> +
> +
> +DOCUMENTATION = '''
> +---
> +module: avi_api
> +author: Gaurav Rastogi (grastogi@avinetworks.com)
> +
> +short_description: Avi API Module.
>
> No fullstop on short_description:
>
Fixed


> ------------------------------
>
> In lib/ansible/modules/network/avi/avi_api_session.py
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20482258>
> :
>
> > +description:
> +    - This module is used to configure Pool object
> +version_added: 2.3
> +requirements: [ avisdk ]
> +options:
> +    http_method:
> +        description:
> +            - Allowed HTTP methods for RESTful services and are supported by Avi Controller.
> +        choices: ["get", "put", "post", "patch", "delete"]
> +        required: true
> +    data:
> +        description:
> +            - HTTP body in YAML format.
> +    data_json:
> +        description:
> +            - HTTP body in JSON format.
>
> It would be good to combine these before the module gets released to avoid
> changing the module interface after release.
>
[GR] We need to keep the current behavior as desired. The problem is type
checking of arguments in the AnsibleModule. If I make the argument type as
dict then I cannot pass in JSON string. If make argument as string then one
cannot pass a dict as the argument.

Eg.  data=dict(type='dict') will fail when json string is passed.
data=dict(type='str') will fail for when data passed down is a dict. So, I
had to split this into two different arguments I we want to support both.


------------------------------
>
> In lib/ansible/modules/network/avi/avi_api_session.py
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20482258>
> :
>
> > +        required: true
> +    data:
> +        description:
> +            - HTTP body in YAML format.
> +    data_json:
> +        description:
> +            - HTTP body in JSON format.
> +    params:
> +        description:
> +            - Query parameters passed to the HTTP API.
> +    path:
> +        description:
> +            - Path for Avi API resource. For example, C(path: virtualservice) will translate to C(api/virtualserivce).
> +    timeout:
> +        description:
> +            - Timeout for Avi API calls.
>
> Timeout (in seconds) for Avi API calls. (assuming it is seconds)
>
fixed

> ------------------------------
>
> In lib/ansible/modules/network/avi/avi_api_session.py
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20482258>
> :
>
> > +            - HTTP body in JSON format.
> +    params:
> +        description:
> +            - Query parameters passed to the HTTP API.
> +    path:
> +        description:
> +            - Path for Avi API resource. For example, C(path: virtualservice) will translate to C(api/virtualserivce).
> +    timeout:
> +        description:
> +            - Timeout for Avi API calls.
> +extends_documentation_fragment:
> +    - avi
> +'''
> +
> +EXAMPLES = '''
> +# Get Pool Information using avi_api_session
>
> Ansible best practice is to specify -name:. This allows people to copy
> and paste the examples to use in their own playbooks. Therefore it's
> clearer to write this as:
>
> - name: Get Pool Information using avi_api_session
>   avi_api_session:
>     controller: "{{ controller }}"
> ...
>
> This applies all examples.
>
fixed

> ------------------------------
>
> In lib/ansible/modules/network/avi/avi_api_session.py
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20482258>
> :
>
> > +    description: Avi REST resource
> +    returned: success, changed
> +    type: dict
> +'''
> +
> +
> +def main():
> +    argument_specs = dict(
> +        http_method=dict(required=True,
> +                         choices=['get', 'put', 'post', 'patch',
> +                                  'delete']),
> +        path=dict(required=True),
> +        params=dict(type='dict'),
> +        data=dict(type='dict'),
> +        data_json=dict(type='str'),
> +        timeout=dict(default=60)
>
> , type='int'
>
fixed

> ------------------------------
>
> In lib/ansible/modules/network/avi/avi_api_session.py
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20482258>
> :
>
> > +    )
> +    argument_specs.update(avi_common_argument_spec())
> +    module = AnsibleModule(argument_spec=argument_specs)
> +
> +    if not HAS_AVI:
> +        return module.fail_json(msg=(
> +            'Avi python API SDK (avisdk) is not installed. '
> +            'For more details visit https://github.com/avinetworks/sdk.'))
> +    tenant_uuid = module.params.get('tenant_uuid', None)
> +    api = ApiSession.get_session(
> +        module.params['controller'], module.params['username'],
> +        module.params['password'], tenant=module.params['tenant'],
> +        tenant_uuid=tenant_uuid)
> +
> +    tenant = module.params.get('tenant', '')
> +    timeout = int(module.params.get('timeout', 60))
>
> You've specified the default here and in argument_specs which doesn't
> seem right
>
fixed

> ------------------------------
>
> In lib/ansible/modules/network/avi/avi_api_session.py
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20482258>
> :
>
> > +options:
> +    http_method:
> +        description:
> +            - Allowed HTTP methods for RESTful services and are supported by Avi Controller.
> +        choices: ["get", "put", "post", "patch", "delete"]
> +        required: true
> +    data:
> +        description:
> +            - HTTP body in YAML format.
> +    data_json:
> +        description:
> +            - HTTP body in JSON format.
> +    params:
> +        description:
> +            - Query parameters passed to the HTTP API.
> +    path:
>
> In argument_specs you've specified that path is required, so that needs
> documenting here.
>
fixed: added comment

> ------------------------------
>
> In lib/ansible/utils/module_docs_fragments/avi.py
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20482258>
> :
>
> > +class ModuleDocFragment(object):
> +    # Avi common documentation fragment
> +    DOCUMENTATION = """
> +options:
> +    controller:
> +        description:
> +            - IP address or hostname of the controller. The default value is the environment variable C(AVI_CONTROLLER).
> +    username:
> +        description:
> +            - Username used for accessing Avi controller. The default value is the environment variable C(AVI_USERNAME).
> +    password:
> +        description:
> +            - Password of Avi user in Avi controller. The default value is the environment variable C(AVI_PASSWORD).
> +    tenant:
> +        description:
> +            - Tenant for the operations
>
> Missing fullstop.
> ------------------------------
>
> In lib/ansible/utils/module_docs_fragments/avi.py
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20482258>
> :
>
> > +    controller:
> +        description:
> +            - IP address or hostname of the controller. The default value is the environment variable C(AVI_CONTROLLER).
> +    username:
> +        description:
> +            - Username used for accessing Avi controller. The default value is the environment variable C(AVI_USERNAME).
> +    password:
> +        description:
> +            - Password of Avi user in Avi controller. The default value is the environment variable C(AVI_PASSWORD).
> +    tenant:
> +        description:
> +            - Tenant for the operations
> +        default: admin
> +    tenant_uuid:
> +        description:
> +            - Tenant uuid for the operations
>
> Missing fullstop.
>
fixed.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/20415#pullrequestreview-20482258>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADrBf_prykLCfTotkt7xg1AOMSdWYzjRks5raFUSgaJpZM4LnQxe>
> .
>



-- 
Gaurav Rastogi

grastogi AT avinetworks.com
Avi Networks

Commit:
fixed pep8 warnings

gundalow commented:
```
2017-02-08 08:38:36 lib/ansible/modules/network/avi/avi_api_session.py
2017-02-08 08:38:36 ============================================================================
2017-02-08 08:38:36 ERROR: DOCUMENTATION.options.path.description.0: expected basestring @ data['options']['path']['description'][0]. Got None
2017-02-08 08:38:36 ERROR: No ANSIBLE_METADATA provided
2017-02-08 08:38:36 ERROR: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA. line 26

```

bcoca commented:
added missing metadata and url to website in description, please make sure any subsequent modules include these.
